lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -789,12 +789,13 @@ def __init__(
         self._jobs = [_sanitize_scrape_configuration(job) for job in jobs]
 
         events = self._charm.on[self._relation_name]
-        self.framework.observe(events.relation_joined, self._set_scrape_metadata)
-        self.framework.observe(events.relation_changed, self._set_scrape_metadata)
+        self.framework.observe(events.relation_joined, self._set_scrape_job_spec)
+        self.framework.observe(events.relation_changed, self._set_scrape_job_spec)
         self.framework.observe(self._service_event, self._set_unit_ip)
+        self.framework.observe(self._charm.on.upgrade_charm, self._set_scrape_job_spec)
 
-    def _set_scrape_metadata(self, event):
-        """Ensure scrape targets metadata is made available to Prometheus.
+    def _set_scrape_job_spec(self, event):
```

Comment:
Would this be eventually called if LeaderElectedEvent fires? Needed for scaledown.

Reply:
There is nothing that needs to be done in particular to handle a leader change. Whoever is the leader will automatically do the right thing.

Reply:
This may not be the case in the following scenario which I observed multiple times with alertmanager:
1. Deploy number of units from the start, such that the leader is not unit/0.
2. Scale down a number of units under the current leader.
3. Units are removed but the new leader is elected only ~10 seconds after scaledown.

Reply:
I still do not see why it should matter. Do elaborate if you see a particular issue ? Do we have a test case ?

Reply:
Looking more thoroughly at the code I see that scrape job spec carry no information about prom units (e.g. IP addresses). In that case it shouldn't matter.